File: src/main/java/rag/app/stages/FinalizeStage.java
package rag.app.stages;

import rag.answer.Answer;
import rag.answer.FallbackHandler;
import rag.app.Context;
import rag.app.StrategyRegistry;
import rag.trace.TraceBus;
import rag.trace.TraceEvent;

public class FinalizeStage implements PipelineStage {

    private final FallbackHandler fallbackHandler;

    public FinalizeStage(StrategyRegistry registry) {
        this.fallbackHandler = registry.getFallbackHandler();
    }

    @Override
    public String getName() {
        return "FinalizeStage";
    }

    @Override
    public void run(Context context, TraceBus traceBus) throws Exception {
        long start = System.currentTimeMillis();
        String error = null;
        boolean usedFallback = false;
        Exception failure = null;

        try {
            Answer answer = context.getAnswer();
            if (answer == null || answer.getText() == null || answer.getText().isBlank()) {
                Answer fallback = fallbackHandler.buildFallback(context);
                context.setAnswer(fallback);
                usedFallback = true;
            } else {
                context.setFallbackReason("AnswerReady");
            }
        } catch (Exception ex) {
            error = ex.getMessage();
            failure = ex;
        } finally {
            long duration = System.currentTimeMillis() - start;
            traceBus.publish(new TraceEvent(
                    getName(),
                    usedFallback ? "fallback=" + context.getFallbackReason() : "answer-ready",
                    duration,
                    error
            ));
        }

        if (failure != null) {
            throw failure;
        }
    }
}

File: src/main/python/rag/app/stages/finalize_stage.py
from typing import Optional
from rag.app.stages.pipeline_stage import PipelineStage
from rag.answer.answer import Answer
from rag.answer.fallback_handler import FallbackHandler
from rag.app.context import Context
from rag.app.strategy_registry import StrategyRegistry
from rag.trace.trace_bus import TraceBus
from rag.trace.trace_event import TraceEvent
from rag.utils.timer_utils import now


class FinalizeStage(PipelineStage):
    def __init__(self, registry: StrategyRegistry) -> None:
        self.fallback_handler: FallbackHandler = registry.get_fallback_handler()

    def get_name(self) -> str:
        return "FinalizeStage"

    def run(self, context: Context, trace_bus: TraceBus) -> None:
        start: int = now()
        error: Optional[str] = None
        used_fallback: bool = False
        failure: Optional[Exception] = None

        try:
            answer: Optional[Answer] = context.get_answer()
            if answer is None or not answer.get_text() or not answer.get_text().strip():
                fallback: Answer = self.fallback_handler.build_fallback(context)
                context.set_answer(fallback)
                used_fallback = True
            else:
                context.set_fallback_reason("AnswerReady")
        except Exception as ex:
            error = str(ex)
            failure = ex
        finally:
            duration: int = now() - start
            summary: str = (
                f"fallback={context.get_fallback_reason()}"
                if used_fallback
                else "answer-ready"
            )
            trace_bus.publish(
                TraceEvent(
                    self.get_name(),
                    summary,
                    duration,
                    error
                )
            )

        if failure is not None:
            raise failure






